[cutlass backend] cache maybe_append_choices

henrylhtsang opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #156781

This PR attempts to cache:
* codegen for cutlass backend for the same kernel. Even if runtime params are different.

From some profiling, most of the time spent is on render. So we only target to cache that part for now. 

The output of render is `code`, and we are able to cache that easily. Also, I have to cache size_args, since it depends on `kernel.get_dynamic_shape_args()`, which depends on the state of self when we call render.

make_key is doing most of the work here: We are hashing on input node layouts, output node layout and op.configuration_name() (this is what hash(op) would do anyway). 

cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @ipiszy @chenyang78 @kadeng @muchulee8 @amjames @chauhang @aakhundov @coconutruben

henrylhtsang committed:
[cutlass backend] cache maybe_append_choices



Differential Revision: [D77274351](https://our.internmc.facebook.com/intern/diff/D77274351/)

[ghstack-poisoned]

laithsakka commented on torch/_inductor/codegen/cuda/cuda_template.py:
```diff
@@ -72,18 +74,70 @@ def supports_epilogue_fusion(op: GemmOperation) -> bool:
         Returns:
             A CUDATemplateCaller object representing the generated CUDA template caller.
         """
+        # Enable caching when generate_with_caching is True and caching is enabled in config
```
this seems way much simpler than the stuff we did for the generate on the TritonTemplate,  ... thinking

laithsakka reviewed (dismissed):
are you trying to cache the generation of the template?
when we did this to triton templates the issue was a bit more than just caching the final output result, unless in your case the same input nodes are passed(not only same size/layouts?)

the output that you are caching includes
```
self.input_nodes,
self.output_node.get_layout(),
```
which are not used in the cache key, and using input_nodes in the cache key probably not what you want. 
I might be missing something but lets pick some time to chat about this tomorrow .

henrylhtsang commented:
> are you trying to cache the generation of the template? when we did this to triton templates the issue was a bit more than just caching the final output result, unless in your case the same input nodes are passed(not only same size/layouts?)
> 
> the output that you are caching includes
> 
> ```
> self.input_nodes,
> self.output_node.get_layout(),
> ```
> 
> which are not used in the cache key, and using input_nodes in the cache key probably not what you want. I might be missing something but lets pick some time to chat about this tomorrow .

am I missing something? I think the cache key depends on 
```
            "layout": str(self.layout),
            "input_layouts": [str(node.get_layout()) for node in self.input_nodes],
```
which should be independent of self.input_nodes right?

laithsakka commented:
> > are you trying to cache the generation of the template? when we did this to triton templates the issue was a bit more than just caching the final output result, unless in your case the same input nodes are passed(not only same size/layouts?)
> > the output that you are caching includes
> > ```
> > self.input_nodes,
> > self.output_node.get_layout(),
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > which are not used in the cache key, and using input_nodes in the cache key probably not what you want. I might be missing something but lets pick some time to chat about this tomorrow .
> 
> am I missing something? I think the cache key depends on
> 
> ```
>             "layout": str(self.layout),
>             "input_layouts": [str(node.get_layout()) for node in self.input_nodes],
> ```
> 
> which should be independent of self.input_nodes right?

right but you are caching
```
return CUDATemplateCaller(
            kernel_name,
            "cutlass_gemm",
            self.input_nodes,
            self.output_node.get_layout(),
            make_kernel_render,
            bmreq,
            supports_epilogue_fusion,
            self,
            kwargs,
            description,
        )
        ```
which includes a ref to the    self.input_nodes themselves, meaning that if you get another input with the same layout you will return CUDATemplateCaller that reference different input_nodes

henrylhtsang committed:
Update on "[cutlass backend] cache maybe_append_choices"


cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx ipiszy chenyang78 kadeng muchulee8 amjames chauhang aakhundov

[ghstack-poisoned]

henrylhtsang committed:
Update on "[cutlass backend] cache maybe_append_choices"


cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx ipiszy chenyang78 kadeng muchulee8 amjames chauhang aakhundov

[ghstack-poisoned]

laithsakka commented on torch/_inductor/codegen/cuda/cuda_template.py:
```diff
@@ -69,30 +74,51 @@ def __init__(
+    def generate_code_and_args(self, **kwargs) -> tuple[str, tuple[int, ...]]:
         """
+        Generate code and args with caching. We cache the code even if runtime
+        args are different.
```
can add more into 
"We cache the code even if runtime args are different."
which is ok because ....

laithsakka commented on torch/_inductor/codegen/cuda/cuda_template.py:
```diff
@@ -69,30 +74,51 @@ def __init__(
-            patch.object(V.graph, "get_dtype", self._fake_get_dtype(self.output_node)),
-            CUDATemplateKernel(
-                kernel_name=kernel_name,
-                runtime_arg_info=self.get_runtime_arg_info(),
```
are we sure that codegen proccess for all kernels does not use runtime_arg_values?
in other words code is the same regardless? since its not part of the cache key. 

if codegen is independent why do we pass it to the CUDATemplateKernel

laithsakka replied:
i see its only used in     def call_kernel(

henrylhtsang replied:
@laithsakka we need `runtime_arg_info=self.get_runtime_arg_info(),` to define the fields of the function in the codegen (here it is const int swizzle), but we don't need the swizzle value (1, 2, 4, 8).

laithsakka replied:
yeh i see, one concern there is no guarantee that any test will break if someone in the future did depend on the value during codegen

laithsakka commented:
I do not have hard blockers for moving forward with this, while i wish things were better organized for soundness, that said I will leave that to more expert reviewers. 

things to consider:
1)  how do we make sure when someone add a new config or update codegen with new dependency on new args the caching is still valid.
2) how to make sure when someone add a dependency on the output of the codegen process caching do not break.
we handle those in template kernel caching by wrapping codegen in a function with explicit args and well defined output
we even have a unit test that would fail if number of inputs to that function changes.

Since i am no export here I will leave the review for inductor folks.

henrylhtsang committed:
Update on "[cutlass backend] cache maybe_append_choices"


This PR attempts to cache:
* codegen for cutlass backend for the same kernel. Even if runtime params are different.

From some profiling, most of the time spent is on render. So we only target to cache that part for now. 

The output of render is `code`, and we are able to cache that easily. Also, I have to cache size_args, since it depends on `kernel.get_dynamic_shape_args()`, which depends on the state of self when we call render.

make_key is doing most of the work here: We are hashing on input node layouts, output node layout and op.configuration_name() (this is what hash(op) would do anyway). 

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx ipiszy chenyang78 kadeng muchulee8 amjames chauhang aakhundov

[ghstack-poisoned]

henrylhtsang commented:
> I do not have hard blockers for moving forward with this, while i wish things were better organized for soundness, that said I will leave that to more expert reviewers.
> 
> things to consider:
> 
> 1. how do we make sure when someone add a new config or update codegen with new dependency on new args the caching is still valid.
> 2. how to make sure when someone add a dependency on the output of the codegen process caching do not break.
>    we handle those in triton template codegen caching by wrapping codegen in a function with explicit args and well defined output
>    we even have a unit test that would fail if number of inputs to that function changes.
> 
> Since i am no expert here I will leave the review for inductor folks.

I guess a priori I don't disagree. But I think not having this cache will block any adoption, which is a bigger problem.

> 1. how do we make sure when someone add a new config or update codegen with new dependency on new args the caching is still valid.

Can you elaborate on what you mean by new config or update codegen?

> 2. how to make sure when someone add a dependency on the output of the codegen process caching do not break.

Same here. Not sure I follow.

henrylhtsang committed:
Update on "[cutlass backend] cache maybe_append_choices"


This PR attempts to cache:
* codegen for cutlass backend for the same kernel. Even if runtime params are different.

From some profiling, most of the time spent is on render. So we only target to cache that part for now. 

The output of render is `code`, and we are able to cache that easily. Also, I have to cache size_args, since it depends on `kernel.get_dynamic_shape_args()`, which depends on the state of self when we call render.

make_key is doing most of the work here: We are hashing on input node layouts, output node layout and op.configuration_name() (this is what hash(op) would do anyway). 

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx ipiszy chenyang78 kadeng muchulee8 amjames chauhang aakhundov

[ghstack-poisoned]

henrylhtsang committed:
Update on "[cutlass backend] cache maybe_append_choices"


This PR attempts to cache:
* codegen for cutlass backend for the same kernel. Even if runtime params are different.

From some profiling, most of the time spent is on render. So we only target to cache that part for now. 

The output of render is `code`, and we are able to cache that easily. Also, I have to cache size_args, since it depends on `kernel.get_dynamic_shape_args()`, which depends on the state of self when we call render.

make_key is doing most of the work here: We are hashing on input node layouts, output node layout and op.configuration_name() (this is what hash(op) would do anyway). 

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx ipiszy chenyang78 kadeng muchulee8 amjames chauhang aakhundov

[ghstack-poisoned]

ColinPeppler commented on torch/_inductor/codegen/cuda/cuda_template.py:
```diff
@@ -38,8 +40,11 @@ class ArgInfo:
+@clear_on_fresh_cache
 class CUDATemplate(KernelTemplate):
     index_counter = itertools.count()
+    code_cache: dict[str, tuple[str, tuple[int, ...]]] = {}
```
commenton what `tuple[str, tuple[int, ...]]` represents would help readers.

ColinPeppler commented on torch/_inductor/codegen/cuda/cuda_template.py:
```diff
@@ -117,8 +143,35 @@ def generate(  # type: ignore[override]
         size_args = V.graph.sizevars.size_hints(kernel.get_dynamic_shape_args())
+
+        if key is not None:
+            self.code_cache[key] = code, size_args
```
this is actually a good point, are we sure it's safe to cache `size_args`, since `size_args` may contain free_symbols?

https://github.com/pytorch/pytorch/blob/a5e68814d556cf67c6511876410970dd08c3dd6d/torch/_inductor/codegen/cuda/cuda_kernel.py#L307

henrylhtsang committed:
Update on "[cutlass backend] cache maybe_append_choices"


This PR attempts to cache:
* codegen for cutlass backend for the same kernel. Even if runtime params are different.

From some profiling, most of the time spent is on render. So we only target to cache that part for now. 

The output of render is `code`, and we are able to cache that easily. Also, I have to cache size_args, since it depends on `kernel.get_dynamic_shape_args()`, which depends on the state of self when we call render.

make_key is doing most of the work here: We are hashing on input node layouts, output node layout and op.configuration_name() (this is what hash(op) would do anyway). 

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx ipiszy chenyang78 kadeng muchulee8 amjames chauhang aakhundov

[ghstack-poisoned]

henrylhtsang commented:
@pytorchbot merge

henrylhtsang commented:
@pytorchbot rebase -b viable/strict

henrylhtsang committed:
Update

[ghstack-poisoned]

drisspg commented on test/inductor/test_cutlass_backend.py:
```diff
@@ -1552,6 +1553,46 @@ def mm(a, b):
+        )
+        total_time_spent = sum(time_spent)
+        self.assertGreater(total_time_spent, 0)
+        self.assertLess(total_time_spent, 3)
```
ditto, what are the units here?

henrylhtsang replied:
unit is number of seconds. 

Let me think a bit

drisspg replied:
maybe use unitttest.mock to just assert a certain call count e.g. render should be called 1 the rest should go through the cache

henrylhtsang committed:
Update on "[cutlass backend] cache maybe_append_choices"


This PR attempts to cache:
* codegen for cutlass backend for the same kernel. Even if runtime params are different.

From some profiling, most of the time spent is on render. So we only target to cache that part for now. 

The output of render is `code`, and we are able to cache that easily. Also, I have to cache size_args, since it depends on `kernel.get_dynamic_shape_args()`, which depends on the state of self when we call render.

make_key is doing most of the work here: We are hashing on input node layouts, output node layout and op.configuration_name() (this is what hash(op) would do anyway). 

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx ipiszy chenyang78 kadeng muchulee8 amjames chauhang aakhundov

[ghstack-poisoned]

mlazos commented on torch/_inductor/codegen/cuda/cuda_template.py:
```diff
@@ -72,18 +74,70 @@ def supports_epilogue_fusion(op: GemmOperation) -> bool:
         self,
         description,
+        generate_with_caching=False,
         **kwargs,
```
I think so (in order to be correct), but might be worth verifying.
